From bb2f08ef0884718da357a1a3b761cefc23f18099 Mon Sep 17 00:00:00 2001
From: Oleg Komarov <oleg@blender.org>
Date: Wed, 28 Jan 2026 12:35:38 +0100
Subject: [PATCH 19/22] BLENDER: contributor agreement support

New models ContributorAgreement and SignedContributorAgreement are
introduced.

A new Settings > Contributor Agreements page is introduced.
The page allows a user to review the text of existing contributor
agreements and to sign any of them.

A new /api/v1/users/{username}/contributor-agreements/{slug} endpoint
allows to check if a user has signed a contributor agreement.
This api endpoint can be used by a gitea action workflow as described in
https://projects.blender.org/infrastructure/clacheck-runner

SignContributorAgreement function searches for failed cla.yml workflows
that were triggered by the current user and reruns them (search limited to
10 latest workflow runs).
This updates the failed check status automatically and doesn't require any
additional actions from the user.

The runner code https://projects.blender.org/infrastructure/clacheck-runner
automatically posts a pull request comment when a check fails:

    @username please sign the contributor agreement $CLA_SLUG at
    https://projects.blender.org/user/settings/contributor_agreements
---
 models/user/contributor_agreement.go          |  62 ++++++
 options/locale/locale_en-US.json              |  20 +-
 routers/api/v1/api.go                         |   2 +
 routers/api/v1/user/contributor_agreement.go  |  83 ++++++++
 .../admin/signed_contributor_agreements.go    | 201 ++++++++++++++++++
 .../user/setting/contributor_agreements.go    | 158 ++++++++++++++
 routers/web/web.go                            |   9 +
 services/user/delete.go                       |   2 +
 templates/admin/navbar.tmpl                   |   5 +-
 .../batch_sign.tmpl                           |  35 +++
 .../signed_contributor_agreements/list.tmpl   |  60 ++++++
 templates/swagger/v1_json.tmpl                |  36 ++++
 .../user/settings/contributor_agreements.tmpl |  45 ++++
 templates/user/settings/navbar.tmpl           |   3 +
 14 files changed, 719 insertions(+), 2 deletions(-)
 create mode 100644 models/user/contributor_agreement.go
 create mode 100644 routers/api/v1/user/contributor_agreement.go
 create mode 100644 routers/web/admin/signed_contributor_agreements.go
 create mode 100644 routers/web/user/setting/contributor_agreements.go
 create mode 100644 templates/admin/signed_contributor_agreements/batch_sign.tmpl
 create mode 100644 templates/admin/signed_contributor_agreements/list.tmpl
 create mode 100644 templates/user/settings/contributor_agreements.tmpl

diff --git a/models/user/contributor_agreement.go b/models/user/contributor_agreement.go
new file mode 100644
index 0000000000..0dd3373fbc
--- /dev/null
+++ b/models/user/contributor_agreement.go
@@ -0,0 +1,62 @@
+// Copyright 2025 The Gitea Authors. All rights reserved.
+// SPDX-License-Identifier: MIT
+
+// BLENDER: contributor agreement
+
+package user
+
+import (
+	"gitea.dev/models/db"
+	"gitea.dev/modules/timeutil"
+
+	"xorm.io/builder"
+)
+
+type ContributorAgreement struct {
+	ID          int64              `xorm:"pk autoincr"`
+	Slug        string             `xorm:"UNIQUE"`
+	CreatedUnix timeutil.TimeStamp `xorm:"created"`
+	Content     string             `xorm:"TEXT NOT NULL"`
+}
+
+type SignedContributorAgreement struct {
+	ID                     int64              `xorm:"pk autoincr"`
+	UserID                 int64              `xorm:"UNIQUE(s)"`
+	ContributorAgreementID int64              `xorm:"UNIQUE(s)"`
+	CreatedUnix            timeutil.TimeStamp `xorm:"created"`
+	Comment                string             `xorm:"TEXT DEFAULT NULL"`
+}
+
+type FindSignedContributorAgreementsOptions struct {
+	db.ListOptions
+	ContributorAgreementID int64
+	UserID                 int64
+	UserIDs                []int64
+}
+
+func (opts *FindSignedContributorAgreementsOptions) ToConds() builder.Cond {
+	cond := builder.NewCond()
+	if opts.ContributorAgreementID != 0 {
+		cond = cond.And(builder.Eq{"contributor_agreement_id": opts.ContributorAgreementID})
+	}
+	if opts.UserID != 0 {
+		cond = cond.And(builder.Eq{"user_id": opts.UserID})
+	}
+	if opts.UserIDs != nil {
+		cond = cond.And(builder.In("user_id", opts.UserIDs))
+	}
+	return cond
+}
+
+func (opts *FindSignedContributorAgreementsOptions) ToOrders() string {
+	return "id DESC"
+}
+
+func init() {
+	// These tables don't exist in the upstream code.
+	// We don't introduce migrations for it to avoid migration id clashes.
+	// Gitea will create the tables in the database during startup,
+	// so no manual action is required until we start modifying the tables.
+	db.RegisterModel(new(ContributorAgreement))
+	db.RegisterModel(new(SignedContributorAgreement))
+}
diff --git a/options/locale/locale_en-US.json b/options/locale/locale_en-US.json
index 0907aeac35..2ddbfd3c46 100644
--- a/options/locale/locale_en-US.json
+++ b/options/locale/locale_en-US.json
@@ -675,6 +675,7 @@
   "settings.organization": "Organizations",
   "settings.uid": "UID",
   "settings.webauthn": "Two-Factor Authentication (Security Keys)",
+  "settings.contributor_agreements": "Contributor Agreements",
   "settings.public_profile": "Public Profile",
   "settings.biography_placeholder": "Tell us a little bit about yourself! (You can use Markdown)",
   "settings.location_placeholder": "Share your approximate location with others",
@@ -2944,6 +2945,7 @@
   "admin.total": "Total: %d",
   "admin.settings": "Admin Settings",
   "admin.spamreports": "Spam Reports",
+  "admin.signed_contributor_agreements": "Signed CLA",
   "admin.dashboard.new_version_hint": "Gitea %s is now available, you are running %s. Check <a target=\"_blank\" rel=\"noreferrer\" href=\"%s\">the blog</a> for more details.",
   "admin.dashboard.statistic": "Summary",
   "admin.dashboard.maintenance_operations": "Maintenance Operations",
@@ -3111,6 +3113,17 @@
   "admin.spamreports.created": "Report Created",
   "admin.spamreports.updated": "Report Updated",
   "admin.spamreports.status": "Report Status",
+  "admin.signed_contributor_agreements.all_option": "All",
+  "admin.signed_contributor_agreements.batch_sign": "Batch sign contributor agreements",
+  "admin.signed_contributor_agreements.batch_sign_button": "Batch sign",
+  "admin.signed_contributor_agreements.comment": "Comment",
+  "admin.signed_contributor_agreements.panel": "Signed Contributor Agreements",
+  "admin.signed_contributor_agreements.search_button": "Search",
+  "admin.signed_contributor_agreements.search_user": "Search users",
+  "admin.signed_contributor_agreements.signed_at": "Signed At",
+  "admin.signed_contributor_agreements.slug": "CLA",
+  "admin.signed_contributor_agreements.user": "User",
+  "admin.signed_contributor_agreements.usernames_input": "Usernames, one per line",
   "admin.orgs.org_manage_panel": "Organization Management",
   "admin.orgs.name": "Name",
   "admin.orgs.teams": "Teams",
@@ -3943,5 +3956,10 @@
   "actions.general.cross_repo_desc": "Allow the selected repositories to be accessed (read-only) by all the repositories in this owner with GITEA_TOKEN when running Actions jobs.",
   "actions.general.cross_repo_selected": "Selected repositories",
   "actions.general.cross_repo_target_repos": "Target Repositories",
-  "actions.general.cross_repo_add": "Add Target Repository"
+  "actions.general.cross_repo_add": "Add Target Repository",
+  "contributor_agreements.signed_at": "Signed by you at %s",
+  "contributor_agreements.confirm": "I agree to the terms of the contributor agreement.",
+  "contributor_agreements.sign": "Sign",
+  "contributor_agreements.view_and_sign": "View text and sign",
+  "contributor_agreements.view": "View text"
 }
diff --git a/routers/api/v1/api.go b/routers/api/v1/api.go
index fa8d29c28b..1b720199c3 100644
--- a/routers/api/v1/api.go
+++ b/routers/api/v1/api.go
@@ -1084,6 +1084,8 @@ func Routes() *web.Router {
 				}, reqSelfOrAdmin(), reqBasicOrRevProxyAuth())
 
 				m.Get("/activities/feeds", user.ListUserActivityFeeds)
+				// BLENDER: contributor agreement
+				m.Get("/contributor-agreements/{slug}", user.CheckContributorAgreement)
 			}, context.UserAssignmentAPI(), checkTokenPublicOnly(), individualPermsChecker)
 		}, tokenRequiresScopes(auth_model.AccessTokenScopeCategoryUser))
 
diff --git a/routers/api/v1/user/contributor_agreement.go b/routers/api/v1/user/contributor_agreement.go
new file mode 100644
index 0000000000..5a75897016
--- /dev/null
+++ b/routers/api/v1/user/contributor_agreement.go
@@ -0,0 +1,83 @@
+// Copyright 2025 The Gitea Authors. All rights reserved.
+// SPDX-License-Identifier: MIT
+
+// BLENDER: contributor agreement
+
+package user
+
+import (
+	"fmt"
+	"net/http"
+	"strings"
+
+	"gitea.dev/models/db"
+	user_model "gitea.dev/models/user"
+	"gitea.dev/modules/httplib"
+	"gitea.dev/modules/setting"
+	"gitea.dev/services/context"
+
+	"xorm.io/builder"
+)
+
+func CheckContributorAgreement(ctx *context.APIContext) {
+	// swagger:operation GET /users/{username}/contributor-agreements/{slug} user userCheckContributorAgreement
+	// ---
+	// summary: Check if contributor agreement is signed by the user
+	// produces:
+	// - application/json
+	// parameters:
+	// - name: username
+	//   in: path
+	//   description: username of user
+	//   type: string
+	//   required: true
+	// - name: slug
+	//   in: path
+	//   description: slug of a contributor agreement to check
+	//   type: string
+	//   required: true
+	// responses:
+	//   "200":
+	//     "$ref": "#/responses/string"
+	//   "404":
+	//     "$ref": "#/responses/notFound"
+
+	slug := ctx.PathParam("slug")
+	ca, exist, err := db.Get[user_model.ContributorAgreement](ctx, builder.Eq{"slug": slug})
+	if err != nil {
+		ctx.APIErrorInternal(err)
+		return
+	}
+	if !exist {
+		ctx.APIError(http.StatusNotFound, fmt.Sprintf("Contributor agreement <%s> is not found.", slug))
+		return
+	}
+	_, exist, err = db.Get[user_model.SignedContributorAgreement](ctx, builder.Eq{
+		"contributor_agreement_id": ca.ID,
+		"user_id":                  ctx.ContextUser.ID,
+	})
+	if err != nil {
+		ctx.APIErrorInternal(err)
+		return
+	}
+	if !exist {
+		url := httplib.MakeAbsoluteURL(ctx, setting.AppSubURL+"/user/settings/contributor_agreements")
+		message := []string{
+			fmt.Sprintf("Contributor agreement <%s> is not signed.", slug),
+			fmt.Sprintf("Sign at %s", url),
+		}
+		// Add a visible *** decoration.
+		maxlen := 0
+		for i := range message {
+			l := len(message[i])
+			if l > maxlen {
+				maxlen = l
+			}
+		}
+		hr := strings.Repeat("*", maxlen)
+		message = append(append([]string{"", "", hr}, message...), hr, "")
+		ctx.PlainText(http.StatusNotFound, strings.Join(message, "\n"))
+		return
+	}
+	ctx.PlainText(http.StatusOK, "OK")
+}
diff --git a/routers/web/admin/signed_contributor_agreements.go b/routers/web/admin/signed_contributor_agreements.go
new file mode 100644
index 0000000000..3f9e760101
--- /dev/null
+++ b/routers/web/admin/signed_contributor_agreements.go
@@ -0,0 +1,201 @@
+// Copyright 2025 The Gitea Authors.
+// SPDX-License-Identifier: MIT
+
+// BLENDER: contributor agreement
+
+package admin
+
+import (
+	"fmt"
+	"net/http"
+	"strings"
+
+	"gitea.dev/models/db"
+	user_model "gitea.dev/models/user"
+	"gitea.dev/modules/log"
+	"gitea.dev/modules/setting"
+	"gitea.dev/modules/templates"
+	"gitea.dev/services/context"
+)
+
+const (
+	tplSignedContributorAgreements    templates.TplName = "admin/signed_contributor_agreements/list"
+	tplContributorAgreementsBatchSign templates.TplName = "admin/signed_contributor_agreements/batch_sign"
+)
+
+// SignedContributorAgreements shows signed contributor agreements.
+func SignedContributorAgreements(ctx *context.Context) {
+	ctx.Data["Title"] = ctx.Tr("admin.signed_contributor_agreements")
+	ctx.Data["PageIsSignedContributorAgreements"] = true
+
+	page := ctx.FormInt("page")
+	if page <= 1 {
+		page = 1
+	}
+
+	opts := user_model.FindSignedContributorAgreementsOptions{
+		ListOptions: db.ListOptions{
+			PageSize: setting.UI.Admin.UserPagingNum,
+			Page:     page,
+		},
+	}
+
+	contributorAgreementID := ctx.FormInt64("contributor_agreement_id")
+	ctx.Data["ContributorAgreementID"] = contributorAgreementID
+	if contributorAgreementID != 0 {
+		opts.ContributorAgreementID = contributorAgreementID
+	}
+
+	keyword := ctx.FormTrim("q")
+	ctx.Data["Keyword"] = keyword
+	if keyword != "" {
+		users, _, err := user_model.SearchUsers(ctx, user_model.SearchUserOptions{Keyword: keyword, SearchByEmail: true})
+		if err != nil {
+			ctx.ServerError("SearchUsers", err)
+			return
+		}
+		userIDs := make([]int64, 0, len(users))
+		for _, user := range users {
+			userIDs = append(userIDs, user.ID)
+		}
+		opts.UserIDs = userIDs
+	}
+
+	signedContributorAgreements, count, err := db.FindAndCount[user_model.SignedContributorAgreement](ctx, &opts)
+	if err != nil {
+		ctx.ServerError("SignedContributorAgreements", err)
+		return
+	}
+	ctx.Data["SignedContributorAgreements"] = signedContributorAgreements
+	pager := context.NewPagination(count, opts.PageSize, opts.Page, 5)
+	pager.AddParamFromRequest(ctx.Req)
+	ctx.Data["Page"] = pager
+
+	contributorAgreements, err := db.Find[user_model.ContributorAgreement](ctx, &db.ListOptions{})
+	if err != nil {
+		ctx.ServerError("FindContributorAgreements", err)
+		return
+	}
+	ctx.Data["ContributorAgreements"] = contributorAgreements
+	contributorAgreementsLookup := make(map[int64]string)
+	for _, ca := range contributorAgreements {
+		contributorAgreementsLookup[ca.ID] = ca.Slug
+	}
+	ctx.Data["ContributorAgreementsLookup"] = contributorAgreementsLookup
+
+	userIDs := make([]int64, 0, len(signedContributorAgreements))
+	for _, sca := range signedContributorAgreements {
+		userIDs = append(userIDs, sca.UserID)
+	}
+	userMap, err := user_model.GetUsersMapByIDs(ctx, userIDs)
+	if err != nil {
+		ctx.ServerError("GetUsersMapByIDs", err)
+		return
+	}
+	ctx.Data["UserMap"] = userMap
+
+	ctx.HTML(http.StatusOK, tplSignedContributorAgreements)
+}
+
+// ContributorAgreementsBatchSign displays a form for batch signing contributor agreements on users' behalf.
+func ContributorAgreementsBatchSign(ctx *context.Context) {
+	ctx.Data["Title"] = ctx.Tr("admin.signed_contributor_agreements.batch_sign")
+	ctx.Data["PageIsSignedContributorAgreements"] = true
+
+	contributorAgreements, err := db.Find[user_model.ContributorAgreement](ctx, &db.ListOptions{})
+	if err != nil {
+		ctx.ServerError("FindContributorAgreements", err)
+		return
+	}
+	ctx.Data["ContributorAgreements"] = contributorAgreements
+
+	ctx.HTML(http.StatusOK, tplContributorAgreementsBatchSign)
+}
+
+// ContributorAgreementsBatchSign processes a form for batch signing contributor agreements on users' behalf.
+func ContributorAgreementsBatchSignPost(ctx *context.Context) {
+	comment := ctx.FormTrim("comment")
+	contributorAgreementID := ctx.FormInt64("contributor_agreement_id")
+	usernames := strings.Fields(ctx.FormTrim("usernames"))
+
+	ctx.Data["Title"] = ctx.Tr("admin.signed_contributor_agreements.batch_sign")
+	ctx.Data["PageIsSignedContributorAgreements"] = true
+
+	ctx.Data["Comment"] = comment
+	ctx.Data["ContributorAgreementID"] = contributorAgreementID
+	ctx.Data["Usernames"] = strings.Join(usernames, "\n")
+
+	contributorAgreements, err := db.Find[user_model.ContributorAgreement](ctx, &db.ListOptions{})
+	if err != nil {
+		ctx.ServerError("FindContributorAgreements", err)
+		return
+	}
+	ctx.Data["ContributorAgreements"] = contributorAgreements
+
+	exists, err := db.ExistByID[user_model.ContributorAgreement](ctx, contributorAgreementID)
+	if err != nil {
+		ctx.ServerError("ExistByID", err)
+		return
+	}
+	if !exists {
+		ctx.Data["Error"] = fmt.Sprintf("unknown contributor agreement ID=%d", contributorAgreementID)
+		ctx.HTML(http.StatusOK, tplContributorAgreementsBatchSign)
+		return
+	}
+
+	userIDs, err := user_model.GetUserIDsByNames(ctx, usernames, false)
+	if err != nil {
+		if user_model.IsErrUserNotExist(err) {
+			ctx.Data["Error"] = err
+			log.Error("GetUserIDsByNames: %v", err)
+		} else {
+			ctx.Data["Error"] = "something went wrong"
+		}
+		ctx.HTML(http.StatusOK, tplContributorAgreementsBatchSign)
+		return
+	}
+	// Check for duplicates before inserting.
+	existingSCA, err := db.Find[user_model.SignedContributorAgreement](ctx, &user_model.FindSignedContributorAgreementsOptions{
+		ContributorAgreementID: contributorAgreementID,
+		UserIDs:                userIDs,
+	})
+	if err != nil {
+		ctx.ServerError("FindSignedContributorAgreements", err)
+		return
+	}
+	if len(existingSCA) > 0 {
+		existingUserIDs := make([]int64, len(existingSCA))
+		for i, sca := range existingSCA {
+			existingUserIDs[i] = sca.UserID
+		}
+		existingUsers, err := user_model.GetUserByIDs(ctx, existingUserIDs)
+		if err != nil {
+			ctx.ServerError("GetUserByIDs", err)
+			return
+		}
+		existingUsernames := make([]string, len(existingUsers))
+		for i, user := range existingUsers {
+			existingUsernames[i] = user.Name
+		}
+		ctx.Data["Error"] = fmt.Sprintf("users already signed this contributor agreement: %v", existingUsernames)
+		ctx.HTML(http.StatusOK, tplContributorAgreementsBatchSign)
+		return
+	}
+
+	insertSCA := make([]*user_model.SignedContributorAgreement, len(userIDs))
+	for i, userID := range userIDs {
+		insertSCA[i] = &user_model.SignedContributorAgreement{
+			ContributorAgreementID: contributorAgreementID,
+			Comment:                comment,
+			UserID:                 userID,
+		}
+	}
+	if err := db.Insert(ctx, insertSCA); err != nil {
+		log.Error("Insert SignedContributorAgreements: %v", err)
+		ctx.Data["Error"] = "something went wrong"
+		ctx.HTML(http.StatusOK, tplContributorAgreementsBatchSign)
+		return
+	}
+
+	ctx.Redirect(setting.AppSubURL + "/-/admin/signed_contributor_agreements")
+}
diff --git a/routers/web/user/setting/contributor_agreements.go b/routers/web/user/setting/contributor_agreements.go
new file mode 100644
index 0000000000..3544e07bea
--- /dev/null
+++ b/routers/web/user/setting/contributor_agreements.go
@@ -0,0 +1,158 @@
+// Copyright 2025 The Gitea Authors. All rights reserved.
+// SPDX-License-Identifier: MIT
+
+// BLENDER: contributor agreement
+
+package setting
+
+import (
+	"context"
+	"fmt"
+	"net/http"
+
+	actions_model "gitea.dev/models/actions"
+	"gitea.dev/models/db"
+	user_model "gitea.dev/models/user"
+	"gitea.dev/modules/log"
+	"gitea.dev/modules/setting"
+	"gitea.dev/modules/templates"
+	"gitea.dev/modules/timeutil"
+	actions_service "gitea.dev/services/actions"
+	gitea_context "gitea.dev/services/context"
+	notify_service "gitea.dev/services/notify"
+
+	"xorm.io/builder"
+)
+
+const (
+	tplSettingsContributorAgreements templates.TplName = "user/settings/contributor_agreements"
+)
+
+// ContributorAgreements lists all agreements and shows which ones were signed.
+func ContributorAgreements(ctx *gitea_context.Context) {
+	contributorAgreements, err := db.Find[user_model.ContributorAgreement](ctx, &db.ListOptions{})
+	if err != nil {
+		ctx.ServerError("FindContributorAgreements", err)
+		return
+	}
+	signedContributorAgreements, err := db.Find[user_model.SignedContributorAgreement](ctx, &user_model.FindSignedContributorAgreementsOptions{UserID: ctx.Doer.ID})
+	if err != nil {
+		ctx.ServerError("FindSignedContributorAgreements", err)
+		return
+	}
+	id2sca := make(map[int64]*user_model.SignedContributorAgreement)
+	for _, sca := range signedContributorAgreements {
+		id2sca[sca.ContributorAgreementID] = sca
+	}
+	type UserContributorAgreement struct {
+		user_model.ContributorAgreement
+		Comment  string
+		SignedAt timeutil.TimeStamp
+	}
+	userContributorAgreements := make([]*UserContributorAgreement, len(contributorAgreements))
+	for i, ca := range contributorAgreements {
+		var uca UserContributorAgreement
+		userContributorAgreements[i] = &uca
+		uca.ContributorAgreement = *ca
+		if sca, has := id2sca[ca.ID]; has {
+			uca.Comment = sca.Comment
+			uca.SignedAt = sca.CreatedUnix
+		}
+	}
+
+	ctx.Data["ContributorAgreements"] = userContributorAgreements
+	ctx.Data["PageIsSettingsContributorAgreements"] = true
+	ctx.Data["Title"] = ctx.Tr("settings.contributor_agreements")
+	ctx.HTML(http.StatusOK, tplSettingsContributorAgreements)
+}
+
+// SignContributorAgreement signs a contributor agreement.
+func SignContributorAgreement(ctx *gitea_context.Context) {
+	slug := ctx.PathParam("slug")
+	ca, exist, err := db.Get[user_model.ContributorAgreement](ctx, builder.Eq{"slug": slug})
+	if err != nil {
+		ctx.ServerError("GetContributorAgreement", err)
+		return
+	}
+	if !exist {
+		ctx.PlainText(http.StatusNotFound, "unknown contributor agreement")
+		return
+	}
+	// Prefer to fail on duplicate submission (e.g. when a user hits the button twice).
+	// If rerunFailedActions fails, we won't allow to re-enter the flow and
+	// - either a manual re-run by an admin
+	// - or a re-trigger by the workflow conditions would be required.
+	sca := user_model.SignedContributorAgreement{UserID: ctx.Doer.ID, ContributorAgreementID: ca.ID}
+	if err := db.Insert(ctx, &sca); err != nil {
+		ctx.ServerError("SignContributorAgreement", err)
+		return
+	}
+	// Check if there are any recent action run failures and trigger a re-run.
+	// Execute all updates transactionally, otherwise an ActionRun may be left in Waiting status indefinitely
+	// when UpdateRun is called but UpdateRunJob is not, e.g. due to request being cancelled by a user.
+	//
+	// Currently it seems better not to couple signing and rerun in a single transaction,
+	// to make sure that nothing interferes with storing of the sca record.
+	// Theoretically a rerun may be processed asynchronously.
+	doer := ctx.Doer
+	if err := db.WithTx(ctx, func(ctx context.Context) error {
+		return rerunFailedActions(ctx, doer)
+	}); err != nil {
+		log.Error("rerunFailedActions for userID=%d failed: %v", doer.ID, err)
+	}
+	ctx.Redirect(setting.AppSubURL + "/user/settings/contributor_agreements")
+}
+
+// rerunFailedActions looks for recent cla.yml failures triggered by the user and reruns those.
+// It doesn't pick up runs that were initiated by someone else, e.g. when synchronizing a PR with a target branch.
+func rerunFailedActions(ctx context.Context, user *user_model.User) error {
+	runs, err := db.Find[actions_model.ActionRun](ctx, actions_model.FindRunOptions{
+		ListOptions:   db.ListOptions{PageSize: 10},
+		Status:        []actions_model.Status{actions_model.StatusFailure},
+		TriggerUserID: user.ID,
+		WorkflowID:    "cla.yml",
+	})
+	if err != nil {
+		return fmt.Errorf("failed to find failed action runs: %w", err)
+	}
+
+	// Core logic copied from Rerun func in routers/web/repo/actions/view.go
+	for _, run := range runs {
+		// reset run's start and stop time
+		run.PreviousDuration = run.Duration()
+		run.Started = 0
+		run.Stopped = 0
+		run.Status = actions_model.StatusWaiting
+		if err := actions_model.UpdateRun(ctx, run, "started", "stopped", "status", "previous_duration"); err != nil {
+			return fmt.Errorf("failed to UpdateRun: %w", err)
+		}
+		if err = run.LoadAttributes(ctx); err != nil {
+			return fmt.Errorf("LoadAttributes: %w", err)
+		}
+		notify_service.WorkflowRunStatusUpdate(ctx, run.Repo, run.TriggerUser, run)
+
+		jobs, err := actions_model.GetRunJobsByRunID(ctx, run.ID)
+		if err != nil {
+			return fmt.Errorf("GetRunJobsByRunID: %w", err)
+		}
+		// We always expect exactly one job in cla.yml
+		if len(jobs) != 1 {
+			return fmt.Errorf("cla.yml has more than one job, run.ID=%d", run.ID)
+		}
+
+		job := jobs[0]
+		status := job.Status
+		job.TaskID = 0
+		job.Status = actions_model.StatusWaiting
+		job.Started = 0
+		job.Stopped = 0
+		if _, err := actions_model.UpdateRunJob(ctx, job, builder.Eq{"status": status}, "task_id", "status", "started", "stopped"); err != nil {
+			return fmt.Errorf("UpdateRunJob failed for job.ID=%d: %w", job.ID, err)
+		}
+
+		actions_service.CreateCommitStatusForRunJobs(ctx, run, job)
+		notify_service.WorkflowJobStatusUpdate(ctx, run.Repo, run.TriggerUser, job, nil)
+	}
+
+	return nil
+}
diff --git a/routers/web/web.go b/routers/web/web.go
index f9d4763016..8239eeafc8 100644
--- a/routers/web/web.go
+++ b/routers/web/web.go
@@ -736,6 +736,9 @@ func registerWebRoutes(m *web.Router, webAuth *AuthMiddleware) {
 
 		// BLENDER: spam reporting
 		m.Post("/spamreport", user_setting.SpamReportUserPost)
+		// BLENDER: contributor agreement
+		m.Get("/contributor_agreements", user_setting.ContributorAgreements)
+		m.Post("/contributor_agreements/{slug}/sign", user_setting.SignContributorAgreement)
 	}, reqSignIn, user_setting.SettingsCtxData)
 
 	m.Group("/user", func() {
@@ -826,6 +829,12 @@ func registerWebRoutes(m *web.Router, webAuth *AuthMiddleware) {
 		})
 		m.Post("/purge_spammer", admin.PurgeSpammerPost)
 
+		// BLENDER: contributor agreement
+		m.Group("/signed_contributor_agreements", func() {
+			m.Get("", admin.SignedContributorAgreements)
+			m.Combo("/batch_sign").Get(admin.ContributorAgreementsBatchSign).Post(admin.ContributorAgreementsBatchSignPost)
+		})
+
 		m.Group("/orgs", func() {
 			m.Get("", admin.Organizations)
 		})
diff --git a/services/user/delete.go b/services/user/delete.go
index 5592ac7c4b..08d5e16f9d 100644
--- a/services/user/delete.go
+++ b/services/user/delete.go
@@ -96,6 +96,8 @@ func deleteUser(ctx context.Context, u *user_model.User, purge bool) (err error)
 		&user_model.Blocking{BlockeeID: u.ID},
 		&actions_model.ActionRunnerToken{OwnerID: u.ID},
 		&actions_model.ActionScopedWorkflowSource{OwnerID: u.ID},
+		// BLENDER: contributor agreement
+		&user_model.SignedContributorAgreement{UserID: u.ID},
 	); err != nil {
 		return fmt.Errorf("deleteBeans: %w", err)
 	}
diff --git a/templates/admin/navbar.tmpl b/templates/admin/navbar.tmpl
index 9268134887..775e29966b 100644
--- a/templates/admin/navbar.tmpl
+++ b/templates/admin/navbar.tmpl
@@ -13,7 +13,7 @@
 				</a>
 			</div>
 		</details>
-		<details class="item" {{if or .PageIsAdminUsers .PageIsAdminBadges .PageIsAdminEmails .PageIsAdminOrganizations .PageIsAdminAuthentications .PageIsSpamReports}}open{{end}}>
+		<details class="item" {{if or .PageIsAdminUsers .PageIsAdminBadges .PageIsAdminEmails .PageIsAdminOrganizations .PageIsAdminAuthentications .PageIsSpamReports .PageIsSignedContributorAgreements}}open{{end}}>
 			<summary>{{ctx.Locale.Tr "admin.identity_access"}}</summary>
 			<div class="menu">
 				<a class="{{if .PageIsAdminAuthentications}}active {{end}}item" href="{{AppSubUrl}}/-/admin/auths">
@@ -34,6 +34,9 @@
 				<a class="{{if .PageIsSpamReports}}active {{end}}item" href="{{AppSubUrl}}/-/admin/spamreports">
 					{{ctx.Locale.Tr "admin.spamreports"}}
 				</a>
+				<a class="{{if .PageIsSignedContributorAgreements}}active {{end}}item" href="{{AppSubUrl}}/-/admin/signed_contributor_agreements">
+					{{ctx.Locale.Tr "admin.signed_contributor_agreements"}}
+				</a>
 			</div>
 		</details>
 		<details class="item" {{if or .PageIsAdminRepositories (and .EnablePackages .PageIsAdminPackages)}}open{{end}}>
diff --git a/templates/admin/signed_contributor_agreements/batch_sign.tmpl b/templates/admin/signed_contributor_agreements/batch_sign.tmpl
new file mode 100644
index 0000000000..9fe5dabc8f
--- /dev/null
+++ b/templates/admin/signed_contributor_agreements/batch_sign.tmpl
@@ -0,0 +1,35 @@
+{{template "admin/layout_head" (dict "ctxData" . "pageClass" "admin user")}}
+
+<div class="admin-setting-content">
+	<h4 class="ui top attached header">
+		{{ctx.Locale.Tr "admin.signed_contributor_agreements.batch_sign"}}
+	</h4>
+	<div class="ui attached segment">
+		{{if .Error}}
+			<div class="ui error message">{{.Error}}</div>
+		{{end}}
+		<form class="ui form" method="post">
+			{{.CsrfTokenHtml}}
+			<label>
+				{{ctx.Locale.Tr "admin.signed_contributor_agreements.slug"}}
+				<select name="contributor_agreement_id" required>
+					<option value="">--</option>
+					{{range .ContributorAgreements}}
+						<option value="{{.ID}}" {{if eq .ID $.ContributorAgreementID}}selected{{end}}>{{.Slug}}</option>
+					{{end}}
+				</select>
+			</label>
+			<label>
+				{{ctx.Locale.Tr "admin.signed_contributor_agreements.usernames_input"}}
+				<textarea name="usernames" required>{{.Usernames}}</textarea>
+			</label>
+			<label>
+				{{ctx.Locale.Tr "admin.signed_contributor_agreements.comment"}}
+				<textarea name="comment" required>{{.Comment}}</textarea>
+			</label>
+			<button class="ui button tw-mt-4" type="submit">{{ctx.Locale.Tr "admin.signed_contributor_agreements.batch_sign_button"}}</button>
+		</form>
+	</div>
+</div>
+
+{{template "admin/layout_footer" .}}
diff --git a/templates/admin/signed_contributor_agreements/list.tmpl b/templates/admin/signed_contributor_agreements/list.tmpl
new file mode 100644
index 0000000000..9c4db6809d
--- /dev/null
+++ b/templates/admin/signed_contributor_agreements/list.tmpl
@@ -0,0 +1,60 @@
+{{template "admin/layout_head" (dict "ctxData" . "pageClass" "admin user")}}
+
+<div class="admin-setting-content">
+	<h4 class="ui top attached header">
+		{{ctx.Locale.Tr "admin.signed_contributor_agreements.panel"}}
+		<div class="ui right">
+			<a class="ui primary tiny button" href="/-/admin/signed_contributor_agreements/batch_sign">
+				{{ctx.Locale.Tr "admin.signed_contributor_agreements.batch_sign_button"}}
+			</a>
+		</div>
+	</h4>
+	<div class="ui attached segment">
+		<form class="ui form ignore-dirty flex-text-block">
+			<label>
+				{{ctx.Locale.Tr "admin.signed_contributor_agreements.search_user"}}
+				<input name="q" value="{{.Keyword}}" />
+			</label>
+			<label>
+				{{ctx.Locale.Tr "admin.signed_contributor_agreements.slug"}}
+				<select name="contributor_agreement_id">
+					<option value="">{{ctx.Locale.Tr "admin.signed_contributor_agreements.all_option"}}</option>
+					{{range .ContributorAgreements}}
+						<option value="{{.ID}}" {{if eq .ID $.ContributorAgreementID}}selected{{end}}>{{.Slug}}</option>
+					{{end}}
+				</select>
+			</label>
+			<button class="ui button tw-mt-4" type="submit">{{ctx.Locale.Tr "admin.signed_contributor_agreements.search_button"}}</button>
+		</form>
+	</div>
+	<div class="ui attached table segment">
+		<table class="ui very basic striped table unstackable">
+			<thead>
+				<tr>
+					<th>{{ctx.Locale.Tr "admin.signed_contributor_agreements.user"}}</th>
+					<th>{{ctx.Locale.Tr "admin.signed_contributor_agreements.signed_at"}}</th>
+					<th>{{ctx.Locale.Tr "admin.signed_contributor_agreements.slug"}}</th>
+					<th>{{ctx.Locale.Tr "admin.signed_contributor_agreements.comment"}}</th>
+				</tr>
+			</thead>
+			<tbody>
+				{{range .SignedContributorAgreements}}
+					<tr>
+						<td>
+							<a href="{{AppSubUrl}}/{{(index $.UserMap .UserID).Name | PathEscape}}">
+								{{(index $.UserMap .UserID).Name}}
+							</a>
+							({{(index $.UserMap .UserID).Email}})
+						</td>
+						<td>{{DateUtils.AbsoluteShort .CreatedUnix}}</td>
+						<td>{{index $.ContributorAgreementsLookup .ContributorAgreementID}}</td>
+						<td>{{.Comment}}</td>
+					</tr>
+				{{end}}
+			</tbody>
+		</table>
+	</div>
+	{{template "base/paginate" .}}
+</div>
+
+{{template "admin/layout_footer" .}}
diff --git a/templates/swagger/v1_json.tmpl b/templates/swagger/v1_json.tmpl
index 41e4209648..c7d864e0bf 100644
--- a/templates/swagger/v1_json.tmpl
+++ b/templates/swagger/v1_json.tmpl
@@ -21397,6 +21397,42 @@
         }
       }
     },
+    "/users/{username}/contributor-agreements/{slug}": {
+      "get": {
+        "produces": [
+          "application/json"
+        ],
+        "tags": [
+          "user"
+        ],
+        "summary": "Check if contributor agreement is signed by the user",
+        "operationId": "userCheckContributorAgreement",
+        "parameters": [
+          {
+            "type": "string",
+            "description": "username of user",
+            "name": "username",
+            "in": "path",
+            "required": true
+          },
+          {
+            "type": "string",
+            "description": "slug of a contributor agreement to check",
+            "name": "slug",
+            "in": "path",
+            "required": true
+          }
+        ],
+        "responses": {
+          "200": {
+            "$ref": "#/responses/string"
+          },
+          "404": {
+            "$ref": "#/responses/notFound"
+          }
+        }
+      }
+    },
     "/users/{username}/followers": {
       "get": {
         "produces": [
diff --git a/templates/user/settings/contributor_agreements.tmpl b/templates/user/settings/contributor_agreements.tmpl
new file mode 100644
index 0000000000..6276dd5b7b
--- /dev/null
+++ b/templates/user/settings/contributor_agreements.tmpl
@@ -0,0 +1,45 @@
+{{template "user/settings/layout_head" (dict "ctxData" . "pageClass" "user settings contributor-agreements")}}
+	<div class="user-setting-content">
+		<h4 class="ui top attached header">
+			{{ctx.Locale.Tr "settings.contributor_agreements"}}
+		</h4>
+		<div class="ui attached segment">
+			<div class="ui list">
+				{{range .ContributorAgreements}}
+					<div class="item">
+						<div class="tw-mb-2">
+							<h4 class="ui header"><strong>{{.Slug}}</strong></h4>
+							{{if .SignedAt}}
+								<div class="text light-3">{{if .Comment}}{{.Comment}}{{else}}{{ctx.Locale.Tr "contributor_agreements.signed_at" (DateUtils.FullTime .SignedAt)}}{{end}}</div>
+							{{end}}
+						</div>
+						<details>
+							<summary>
+								{{if not .SignedAt}}
+									{{ctx.Locale.Tr "contributor_agreements.view_and_sign"}}
+								{{else}}
+									{{ctx.Locale.Tr "contributor_agreements.view"}}
+								{{end}}
+							</summary>
+							<pre style="white-space: pre-wrap; font-family: inherit">{{.Content}}</pre>
+							{{if not .SignedAt}}
+								<form method="post" action="{{AppSubUrl}}/user/settings/contributor_agreements/{{.Slug}}/sign" class="ui form">
+									{{$.CsrfTokenHtml}}
+									<div class="field">
+										<div class="ui checkbox">
+											<label>{{ctx.Locale.Tr "contributor_agreements.confirm"}}</label>
+											<input type="checkbox" required name="confirm" />
+										</div>
+									</div>
+									<button class="ui primary button" type="submit">
+										{{ctx.Locale.Tr "contributor_agreements.sign"}}
+									</button>
+								</form>
+							{{end}}
+						</details>
+					</div>
+				{{end}}
+			</div>
+		</div>
+	</div>
+{{template "user/settings/layout_footer" .}}
diff --git a/templates/user/settings/navbar.tmpl b/templates/user/settings/navbar.tmpl
index c278105243..e15b8712e5 100644
--- a/templates/user/settings/navbar.tmpl
+++ b/templates/user/settings/navbar.tmpl
@@ -71,5 +71,8 @@
 		<a class="{{if .PageIsSettingsRepos}}active {{end}}item" href="{{AppSubUrl}}/user/settings/repos">
 			{{ctx.Locale.Tr "settings.repos"}}
 		</a>
+		<a class="{{if .PageIsSettingsContributorAgreements}}active {{end}}item" href="{{AppSubUrl}}/user/settings/contributor_agreements">
+			{{ctx.Locale.Tr "settings.contributor_agreements"}}
+		</a>
 	</div>
 </div>
-- 
2.53.0

